docs/technical/git-submodules.md

```diff
@@ -0,0 +1,167 @@
+objects in the submodule. This requires a way to store submodules in a jj
+repository that supports readonly operations.
+
+#### Outcomes
+
+- Submodules can be cloned anew
+- New submodule commits can be fetched
+- Submodule history and branches can be viewed
+- Submodule contents are populated in the working copy
+- Superproject gitlink can be updated to an existing submodule commit
+- Conflicts in the superproject gitlink can be resolved to an existing submodule
+  commit
+
+### Phase 2: Snapshotting new changes
+
+This allows a user to write new contents to a submodule and its remote.
+
+#### Outcomes
+
+- Changes in the working copy can be recorded in a submodule commit
+- Submodule branches can be modified
+- Submodules and their branches can be pushed to their remote
+
+### Phase 3: Merging/rebasing/conflicts
+
+This allows merging and rebasing of superproject commits in a content-aware way
+(in contrast to Git, where only the gitlink commit ids are compared), as well as
+workflows that make resolving conflicts easy and sensible.
+
+This can be done in tandem with Phase 2, but will likely require a significant
+amount of design work on its own.
+
+#### Outcomes
+
+- Merged/rebased submodules result in merged/rebased working copy content
+- Merged/rebased working copy content can be committed, possibly by creating
+  sensible merged/rebased submodule commits
+- Merge/rebase between submodule and non-submodule gives a sensible result
+- Merge/rebase between submodule A and submodule B gives a sensible result
+
+### Phase ?: An ideal world
+
+I.e. outcomes we would like to see if there were no constraints whatsoever.
+
+- Rewriting submodule commits rewrites descendants correctly and updates
+  superproject gitlinks.
+- Submodule conflicts automatically resolve to the 'correct' submodule commits,
+  e.g. a merge between superproject commits creating a merge of the submodule
+  commits.
+- Nested submodules are as easy to work with as non-nested submodules.
+
+## Design
+
+### Guiding principles
+
+TODO
+
+### On-disk representation
+
+In the Git backend, submodules will be stored in Git's "absorbed submodule"
+format, i.e. a full Git repository with a Git directory in
+`<superproject-git-dir>/modules/<submodule-name>`, and a working tree with a
+`.git` file. Using Git's own format is the simplest option, even in a
+non-colocated repository, since most (if not all) Git implementations can only
+fetch/push from a Git repository, so fetching and pushing submodules requires
+submodules to be full Git repositories.
+
+Note that submodules merely live as full Git repositories in the Git backend,
+but are not be full jj repositories. This will allow us to create a global view
+over the superproject and submodules, which should allow us to avoid some of
+Git's pitfalls (e.g. there is no obvious way to tell the submodule to preserve
+commits because they are needed by the superproject). Compared to Git, this
```

Comment:
is this is a concern for JJ ? JJ already has refs/jj/keep to keep things alive. 

I have been mulling a bit over this too, and actually thought it would be easier to have a `.jj/modules/$NAME`, which contains a JJ repository. Since Git isn't run directly on the git repo within the jj repo, is there a benefit to keeping the submodules in the git repo?

By contrast, if you use the recursive solution, it means you can get JJ to work on the inner repos for free.

Reply:
> is this is a concern for JJ ? JJ already has refs/jj/keep to keep things alive.

Yeah, gc probably won't be a problem for jj submodules for this reason.

> I have been mulling a bit over this too, and actually thought it would be easier to have a .jj/modules/$NAME, which contains a JJ repository.

I think 'global view' of the repo is valuable for other reasons mostly. @martinvonz brought up the case of the operation log/`jj undo`. Having a single operation log track changes in the submodule and superproject would probably simplify how we implement this vs coordinating between the operation logs of the jj superproject and jj submodule.

I expect a trickle of unknown unknown reasons why we need to coordinate between the superproject and submodule, so I'd still prefer to have the superproject as the entrypoint for everything. That said, maybe it's not so hard to have full jj repos and coordinate between them. I'm not so sure either way, WDYT?

> Since Git isn't run directly on the git repo within the jj repo, is there a benefit to keeping the submodules in the git repo?

Even in a jj-only repository, we do run some Git-like operations in the Git repo (e.g. fetch, push). In _theory_ these could have some smart integrations with submodules, which we might want for `jj git [fetch|push]`. In practice, I don't think these integrations exist (not for `git/git` at least), and I don't know what libgit2 (or the Rust bindings) offer us in this regard (probably not much).

> By contrast, if you use the recursive solution, it means you can get JJ to work on the inner repos for free.

True

Reply:
> Even in a jj-only repository, we do run some Git-like operations in the Git repo (e.g. fetch, push). In theory these could have some smart integrations with submodules, which we might want for jj git [fetch|push]. In practice, I don't think these integrations exist (not for git/git at least), and I don't know what libgit2 (or the Rust bindings) offer us in this regard (probably not much).

My experience is that you don't want your base library (libgit2 in this case) to have 'smart' features. What git/git thinks is a sensible default will probably get in the way in a project that uses git in a different way. If JJ knows what submodules there are, it can coordinate operations across the submodules. The storage layer/API doesn't permit cross-repo operations, so I don't think git/git can essentially do better.

Reply:
> Isn't it too early to make a call here? Ideally, the decision follows from requirements.

Yes, I think you're right. Based on informal discussions, I thought that storing submodules in the Git backend would be uncontroversial, but you've provided good reasons to question that, so I'll step back from this proposal for now.

> it would be great if we could somehow do some experimentation to see what workflow works well in practice, before committing to a decision.

Yes, I'm trying to think of what experiment would be the most educational; I'll probably send another round of RFC to get ideas. Perhaps something like `jj op log` across superproject and submodules would be the most informative, but we'd have to invest substantially into submodules before we could meaningfully test that.

Alternatively, perhaps the better approach is just whichever is less invasive, so that it's easier to undo in case we decide the other approach is better.

Reply:
I'm also unsure if the top-level repo's operation log should record the information about submodules' view objects (refs etc.) or if that should be tracked separately for each submodule and the top-level should just track the operation ids (or view ids?) of the submodules. It might be easiest to implement it as full recursive repos (as @hanwen suggested). We might still want to ignore what the submodules' think is the latest operation according to their "op heads store", though. The superproject should probably be considered the source of truth for that.

Reply:
FWIW, I've been toying with a FUSE filesystem with embedded version control command (inspired by JJ), which implements the full recursive solution.  I haven't gotten round to implementing merge, but might be useful to test some of these options. It has no users, so there is no compatibility to worry about :-)

Reply:
> We might still want to ignore what the submodules' think is the latest operation according to their "op heads store", though. The superproject should probably be considered the source of truth for that.

That's quite promising. I think that would address a major problem with Git + submodules (that I was also trying to eliminate with the 'single operation log' idea), which is that there's no way to transactionally do anything across the superproject and submodules (e.g. if the command succeeds in sub1 but fails in sub2, there isn't a good way to roll back sub1) because each repo is independent and everything is just done via subprocesses. If we take your suggestion where the superproject's operation log tracks the submodule's operation, the superproject can just roll back failed commands by setting the submodule's latest operation.